Cache-Hash-Data-Structure.adoc

```diff
@@ -0,0 +1,209 @@
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
+the configuration.
+
+We follow a state-based CRDT approach, merging the local Hash with the server's Hash. Since the `Hash<?, ?>` structures
+occupy a single entry in the Cache, the whole structure has a version instead of versioning per Hash property. Taking
+advantage of this, we can use the `compute*` method and directly apply the local changes. If a concurrent change happens,
+the `Hash<?, ?>` version is updated, and the replace fails, where the `compute*` retries again until it succeeds. Note
+this could lead to flooding the network, transmitting the `Hash<?, ?>` back and forth, but this only happens with heavy
+concurrent changes over the same structure.
+
+[IMPORTANT]
+.Inconsistency!
+====
+The use of `Hash<?, ?>` could return an outdated value, since the structure is a local copy.
+====
+
+To interact with the `Hash<K, V>`:
+
+[source, java]
+.HashCache.java
+----
+public interface HashCache<K> {
+
+   <HK, HV> HV put(K key, HK hashKey, HV hashValue);
+
+   <HK, HV> Hash<HK, HV> put(K key, Map<HK, HV> values);
+
+   <HK, HV> Map<HK, HV> get(K key, HK hashKey, HK ... others); // <1>
+
+   Hash<HK, HV> get(K key);
+
+   boolean remove(K key);
+
+   <HK> boolean remove(K key, HK hashKey, HK ... others);
+
+   boolean exists(K key);
+
+   <HK> boolean containsKey(K key, HK hashKey);
+
+   <HK, HV> Hash<HK, HV> submit(Hash<HK, HV> hash); // <2>
+}
+----
+
+<1> Returning a `Map<HK, HV>` to contain all the vargs keys.
+<2> The `submit` method is used to apply the local changes to the server's Hash. The server's `Hash<?, ?>` is returned.
+
+I see two options for the operation applying directly to a `Hash<?, ?>`, like `put` and `get`:
+
+1. The interface is just a convenient method. The concrete implementation internally calls the `submit` method,
+effectively transferring the `Hash<?, ?>` to the client, applying the operation, and submitting it to the server.
+2. We implement custom commands for these operations and rely on `compute*` only for the `submit` method. Although,
+unsure if this is feasible because the core would need to know about the `Hash<?, ?>` structure.
+
+And as of Multimap, we provide APIs for:
+
+=== Hot Rod API
+
+Developed only for the new API. Provide implementations for an `AsyncHashCache` and `MutinyHashCache`.
+
+=== REST API
+
+Hash REST API should be the almost the same as the Cache one.
+The difference is in the endpoint.
+
+Examples:
+[source]
+----
+GET /rest/v2/hashes/
+
+GET /rest/v2/hashes/{hashName}?action=???
+
+DELETE /rest/v2/hashes/{hashName}/{hashKey}
+
+DELETE /rest/v2/hashes/{hashName}/{hashKey}/{hashProperty}
+----
+
+* CRUD hashes (providing a configuration to create a hash, templates not supported)
+* Display Hash content entries
+* Get values by key
+* In Indexing is enabled, hashes support query
```

Comment:
How would it do this? Are we planning on having the hashes have their own key and value schema definitions? I don't see that mentioned.

Reply:
This one is straight up a stretch, and one thing I had doubts about (the last bullet point at the end of the doc). If feasible, we would have something similar to the [JSON support](https://redis.io/docs/stack/json/). Should I mark it optional? Or remove?